File: Session_02_Recurrences/multi_language_implementations/recurrence_analysis.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func printIndent(depth int) {
	for i := 0; i < depth; i++ {
		fmt.Printf("  |  ")
	}
}

func calculateFactorial(n int, depth int) int64 {
	printIndent(depth)
	fmt.Printf("Entering factorial(%d). Address of n: %p\n", n, &n)
	if n <= 1 {
		printIndent(depth)
		fmt.Printf("Base Case reached for factorial(%d). Returning 1.\n", n)
		return 1
	}
	subResult := calculateFactorial(n-1, depth+1)
	result := int64(n) * subResult
	printIndent(depth)
	fmt.Printf("Exiting factorial(%d) -> returning %d * %d = %d\n", n, n, subResult, result)
	return result
}

func calculateFibonacci(n int, depth int, branch string) int {
	printIndent(depth)
	fmt.Printf("Entering fibonacci(%d) [%s branch]. Address of n: %p\n", n, branch, &n)
	if n == 0 {
		printIndent(depth)
		fmt.Printf("Base Case: fibonacci(0) = 0. Returning.\n")
		return 0
	}
	if n == 1 {
		printIndent(depth)
		fmt.Printf("Base Case: fibonacci(1) = 1. Returning.\n")
		return 1
	}
	left := calculateFibonacci(n-1, depth+1, "Left")
	right := calculateFibonacci(n-2, depth+1, "Right")
	result := left + right
	printIndent(depth)
	fmt.Printf("Exiting fibonacci(%d) -> returning %d + %d = %d\n", n, left, right, result)
	return result
}

func main() {
	val := 5
	fmt.Printf("=== Recursion & Stack Frame Visualizer ===\n\n")
	fmt.Printf("--- PART 1: Linear Recursion - Factorial(%d) ---\n", val)
	fact := calculateFactorial(val, 0)
	fmt.Printf("\nFactorial of %d is: %d\n\n", val, fact)
	fmt.Printf("--------------------------------------------------\n\n")
	fmt.Printf("--- PART 2: Tree Recursion - Fibonacci(4) ---\n")
	fib := calculateFibonacci(4, 0, "Root")
	fmt.Printf("\nFibonacci of 4 is: %d\n\n", fib)
	fmt.Printf("Press Enter to exit...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
}

File: Session_02_Recurrences/multi_language_implementations/recurrence_analysis.cpp
#include <iostream>
#include <string>

void printIndent(int depth) {
    for (int i = 0; i < depth; i++) {
        std::cout << "  |  ";
    }
}

long long calculateFactorial(int n, int depth) {
    printIndent(depth);
    std::cout << "Entering factorial(" << n << "). Address of n: " << (void*)&n << "\n";
    if (n <= 1) {
        printIndent(depth);
        std::cout << "Base Case reached for factorial(" << n << "). Returning 1.\n";
        return 1;
    }
    long long subResult = calculateFactorial(n - 1, depth + 1);
    long long result = n * subResult;
    printIndent(depth);
    std::cout << "Exiting factorial(" << n << ") -> returning " << n << " * " << subResult << " = " << result << "\n";
    return result;
}

int calculateFibonacci(int n, int depth, const std::string &branch) {
    printIndent(depth);
    std::cout << "Entering fibonacci(" << n << ") [" << branch << " branch]. Address of n: " << (void*)&n << "\n";
    if (n == 0) {
        printIndent(depth);
        std::cout << "Base Case: fibonacci(0) = 0. Returning.\n";
        return 0;
    }
    if (n == 1) {
        printIndent(depth);
        std::cout << "Base Case: fibonacci(1) = 1. Returning.\n";
        return 1;
    }
    int left = calculateFibonacci(n - 1, depth + 1, "Left");
    int right = calculateFibonacci(n - 2, depth + 1, "Right");
    int result = left + right;
    printIndent(depth);
    std::cout << "Exiting fibonacci(" << n << ") -> returning " << left << " + " << right << " = " << result << "\n";
    return result;
}

int main() {
    int val = 5;
    std::cout << "=== Recursion & Stack Frame Visualizer ===\n\n";
    std::cout << "--- PART 1: Linear Recursion - Factorial(" << val << ") ---\n";
    long long fact = calculateFactorial(val, 0);
    std::cout << "\nFactorial of " << val << " is: " << fact << "\n\n";
    std::cout << "--------------------------------------------------\n\n";
    std::cout << "--- PART 2: Tree Recursion - Fibonacci(4) ---\n";
    int fib = calculateFibonacci(4, 0, "Root");
    std::cout << "\nFibonacci of 4 is: " << fib << "\n\n";
    std::cout << "Press Enter to exit...";
    std::cin.get();
    return 0;
}
